lib/AST/ASTImporter.cpp

```diff
@@ -2575,9 +2584,9 @@ ExpectedDecl ASTNodeImporter::VisitRecordDecl(RecordDecl *D) {
     D2 = D2CXX;
     D2->setAccess(D->getAccess());
     D2->setLexicalDeclContext(LexicalDC);
-    if (D->getFriendObjectKind() == Decl::FOK_None &&
-        (!DCXX->getDescribedClassTemplate() || DCXX->isImplicit()))
-      LexicalDC->addDeclInternal(D2);
+    addDeclToDCIfVisible(D, D2, DC, LexicalDC);
+    if (D->getFriendObjectKind() == Decl::FOK_Undeclared)
+      D2->getDeclContext()->getPrimaryContext()->makeDeclVisibleInContext(D2);
```

Comment:
Why do we have to set this? This call enables `D2` to be found by normal C/C++ lookup, but has no effect on the ASTImporterLookupTable (that will contain that Decl anyway).

On the other hand, it makes sense to make the new Decl visible, but only if it was visible in the "From" context. I don't know how to check that (other than doing a normal C++ lookup in the DC).
Also, if we want to handle this, I think it would be the best to do this in Import(Decl*) and we should have tests which check explicitly that the normal C/C++ lookup can find the Decl before and after the import. But this lookup thing should go in a new PR, because this is quite independent from this.
Remember, in the past we removed all calls to `makeDeclVisibleInContext` (https://github.com/Ericsson/clang/pull/566) because we abused it since before we hadn't have the importer specific lookup.

Reply:
I suspect that the friend class (`Y` in the added test case) should be visible somehow in its parent context (but not by adding it). The `class Y` is special in that case because it is not visible in the lexical and normal context but still exists somewhere (if the `Y` is declared before it has its place). 
If this call is not here the test ASTMerge/class2 fails because access to a private variable from a friend class is reported as error.

Reply:
Hmm, I am still not happy with it. Perhaps we should indeed try to fix it in `Import(Decl*)` by doing a `DeclContext::lookup` in the "From" DC and if it is successful only then should we call `makeDeclVisibleInContext` on the "To".
Could you please try if that works?
If it does, then we should have two PRs:
1) Call ToDC->`makeDeclVisibleInContext` if `D` is found by normal lookup in the "from" context.
2) This PR about `addDeclToDCIfVisible`

Reply:
The following should be done:
- If the decl is contained within its parent contexts (normal and lexical), add it by `addDeclInternal`.
- If it is visible (in the "lookup parent", `getLookupParent`?), add it by `makeDeclVisibleInContext`. But only if the previous call was not done (it was not added to any context).
- In minimal import case, at least the `CXXRecordDecl` must be added always to the parent context (even if the From context does not contain it). This is required for LLDB to work: It changes the decl context of the decl (but does not add it to the new one) before calling `Import`, but seems to expect that it is then added (to the new imported context).
- These steps should be done for every `Decl`? If yes this can be done generally by `GetOrCreateImportedDecl` or in `Import(Decl *)`.

Reply:
This sounds reasonable, but there are some questions.

> But only if the previous call was not done (it was not added to any context).

I believe we can have something in the LookupPtr even if that is not contained by the DC, but i cannot really come up with an example. So I think we should just simply call `makeDeclVisibleInContext` if the Decl was visible in the "From" DC.

> In minimal import case, at least the `CXXRecordDecl` must be added always to the parent context

But that is not good, just think about the friend classes. Perhaps we should investigate why LLDB needs this.

> These steps should be done for every `Decl`? If yes this can be done generally by `GetOrCreateImportedDecl` or in `Import(Decl *)`.

I think yes, we should do that.

Reply:
For LLDB case: I meant not totally always, only if it is not a friend and not "templated". Probably otherwise LLDB must be changed to manually add the decl or search for it other way.

Reply:
If it is not a problem that a Decl is added to its context only later after create (more import is possible before the add) it is possible that we add it only in `ImportDeclContext`. This can be somewhat more efficient because (for the imported decls) there is one less remove.